Implement Path::definitelyEqual()

smfr opened the pull request:
#### 4f3c53d350112f5964409a79f655bd3cfc05390b
<pre>
Implement Path::definitelyEqual()
<a href="https://bugs.webkit.org/show_bug.cgi?id=278751">https://bugs.webkit.org/show_bug.cgi?id=278751</a>
<a href="https://rdar.apple.com/134817798">rdar://134817798</a>

Reviewed by Cameron McCormack and Nikolas Zimmermann.

Implement `Path::definitelyEqual()` which will return true if two paths are known to contain
equivalent data. It may return false in some cases for equivalent paths, for example
if one contains a platform path, and one contains a PathStream.

`Path::definitelyEqual()` has to take care to handle the three states of the variant
on both sides.

PathCG and PathSkia can test for platform path equality. PathCairo cannot.

Also implement Path::singleRect() and Path::singleRoundedRect() which will be used
in future.

Add some API tests to exercise Path::definitelyEqual() for various Path states.

* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::Path::definitelyEqual const):
(WebCore::Path::ensureImplForTesting):
(WebCore::Path::singleRect const):
(WebCore::Path::singleRoundedRect const):
* Source/WebCore/platform/graphics/Path.h:
* Source/WebCore/platform/graphics/PathImpl.h:
(WebCore::PathImpl::singleRect const):
(WebCore::PathImpl::singleRoundedRect const):
* Source/WebCore/platform/graphics/PathStream.cpp:
(WebCore::PathStream::definitelyEqual const):
(WebCore::PathStream::singleRect const):
(WebCore::PathStream::singleRoundedRect const):
* Source/WebCore/platform/graphics/PathStream.h:
* Source/WebCore/platform/graphics/cairo/PathCairo.cpp:
(WebCore::PathCairo::definitelyEqual const):
* Source/WebCore/platform/graphics/cairo/PathCairo.h:
* Source/WebCore/platform/graphics/cg/PathCG.cpp:
(WebCore::PathCG::definitelyEqual const):
* Source/WebCore/platform/graphics/cg/PathCG.h:
* Source/WebCore/platform/graphics/skia/PathSkia.cpp:
(WebCore::PathSkia::definitelyEqual const):
* Source/WebCore/platform/graphics/skia/PathSkia.h:
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/PathTests.cpp: Added.
(TestWebKitAPI::TEST(Path, DefinitelyEqual)):
(TestWebKitAPI::TEST(Path, DefinitelyNotEqual)):

Canonical link: <a href="https://commits.webkit.org/282873@main">https://commits.webkit.org/282873@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/75a38913cedf8d8dcb5d9b5ce192be418e8bed1b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64512 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/43878 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17109 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68535 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15119 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/66631 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/51617 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15399 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/51912 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10434 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/67580 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40585 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55831 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32532 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37253 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13209 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13993 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59197 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13538 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70234 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8459 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13045 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59238 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8493 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55920 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/59410 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14231 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7001 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/677 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/39690 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/40768 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/41951 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40511 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebCore/platform/graphics/Path.cpp:
```diff
@@ -364,6 +378,32 @@ std::optional<PathDataLine> Path::singleDataLine() const
     return std::nullopt;
 }
 
+std::optional<PathRect> Path::singleRect() const
```
Ideally these would've been in a separate patch.

heycam commented on Source/WebCore/platform/graphics/Path.h:
```diff
@@ -55,6 +55,8 @@ class Path {
     Path& operator=(const Path&) = default;
     Path& operator=(Path&&) = default;
 
+    bool operator==(const Path&) const;
```
Let's use a name like `isDefinitelyEqual` so that it's clear that we can still return false for some paths that are logically equal (because we don't want to go to the effort of computing something that's comparable, or it's not possible like for Cairo paths).

nikolaszimmermann reviewed (approved):
Looks fine - I would prefer isDefinitelyEqual though.

smfr committed:
Implement Path::definitelyEqual()
https://bugs.webkit.org/show_bug.cgi?id=278751
rdar://134817798

Reviewed by Cameron McCormack and Nikolas Zimmermann.

Implement `Path::definitelyEqual()` which will return true if two paths are known to contain
equivalent data. It may return false in some cases for equivalent paths, for example
if one contains a platform path, and one contains a PathStream.

`Path::definitelyEqual()` has to take care to handle the three states of the variant
on both sides.

PathCG and PathSkia can test for platform path equality. PathCairo cannot.

Also implement Path::singleRect() and Path::singleRoundedRect() which will be used
in future.

Add some API tests to exercise Path::definitelyEqual() for various Path states.

* Source/WebCore/platform/graphics/Path.cpp:
(WebCore::Path::definitelyEqual const):
(WebCore::Path::ensureImplForTesting):
(WebCore::Path::singleRect const):
(WebCore::Path::singleRoundedRect const):
* Source/WebCore/platform/graphics/Path.h:
* Source/WebCore/platform/graphics/PathImpl.h:
(WebCore::PathImpl::singleRect const):
(WebCore::PathImpl::singleRoundedRect const):
* Source/WebCore/platform/graphics/PathStream.cpp:
(WebCore::PathStream::definitelyEqual const):
(WebCore::PathStream::singleRect const):
(WebCore::PathStream::singleRoundedRect const):
* Source/WebCore/platform/graphics/PathStream.h:
* Source/WebCore/platform/graphics/cairo/PathCairo.cpp:
(WebCore::PathCairo::definitelyEqual const):
* Source/WebCore/platform/graphics/cairo/PathCairo.h:
* Source/WebCore/platform/graphics/cg/PathCG.cpp:
(WebCore::PathCG::definitelyEqual const):
* Source/WebCore/platform/graphics/cg/PathCG.h:
* Source/WebCore/platform/graphics/skia/PathSkia.cpp:
(WebCore::PathSkia::definitelyEqual const):
* Source/WebCore/platform/graphics/skia/PathSkia.h:
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/PathTests.cpp: Added.
(TestWebKitAPI::TEST(Path, DefinitelyEqual)):
(TestWebKitAPI::TEST(Path, DefinitelyNotEqual)):

Canonical link: https://commits.webkit.org/282873@main

heycam commented on Source/WebCore/platform/graphics/Path.cpp:
```diff
@@ -65,6 +65,32 @@ Path::Path(const Path& other)
+                return otherSegment && singleSegment == otherSegment.value();
+            }
+
+            return impl.ptr() && other.asImpl() && impl->definitelyEqual(*other.asImpl());
```
Already landed, so not important, but the other functions in this file tend to use the accessor functions rather than WTF::switchOn so if you wanted it to be consistent with those you could write it as:

```
if (isEmpty())
    return other.isEmpty();

if (auto segment = asSingle()) {
    auto otherSegment = other.singleSegment();
    return otherSegment && segment == *otherSegment;
}

if (auto impl = asImpl()) {
    // no need to check for isEmpty() here, since the first branch now checks for this.
    if (auto singleSegment = impl->singleSegment()) {
        auto otherSegment = other.singleSegment();
        return otherSegment && singleSegment == *otherSegment;
    }

    // no need to check for impl.ptr() since impl is a DataRef and it never returns null
    return other.asImpl() && impl->definitelyEqual(*other.asImpl());
}
```
